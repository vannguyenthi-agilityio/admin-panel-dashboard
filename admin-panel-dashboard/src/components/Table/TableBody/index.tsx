import { memo } from "react";

// Types
import { IColumnType } from "@/types";

// Helpers
import { getObjectValue } from "@/helpers";

// Constants
import { RESULT_NOT_FOUND } from "@/constants";

interface IDataTableBody<T> {
  data: T[];
  columns: IColumnType<T>[];
}

const TableBody = <T,>({ data, columns }: IDataTableBody<T>) => {
  return (
    <tbody>
      {data.length ? (
        data.map((item, index )=> {
          const id = getObjectValue(item, "id");
          return (
            <tr key={`table-body-${id}`}>
              {columns.map(column => (
                <td
                  key={`table-row-cell-${column.key}`}
                  className={`px-6 py-2 border-0 border-b border-gray-100 ${index % 2 ? "bg-transparent" : "bg-greyish"}`}>
                  {column.customNode ? (
                    column.customNode(column, item)
                  ) : (
                    <div className="flex items-center">
                      {column.key === "status" && (getObjectValue(item, column.key) === "Approved" ?  <div className="h-2.5 w-2.5 rounded-full bg-green-500 me-2" /> :
                      getObjectValue(item, column.key) === "Rejected" ? <div className="h-2.5 w-2.5 rounded-full bg-red-500 me-2" /> : <div className="h-2.5 w-2.5 rounded-full bg-gray-500 me-2" />
                        )
                      }
                      <p className="text-darker font-normal text-sm">
                        {getObjectValue(item, column.key)}
                      </p>
                    </div>
                  )}
                </td>
              ))}
            </tr>
          );
        })
      ) : (
        <tr className="w-full">
          <td colSpan={10} className="h-32 text-center">
            <p className="text-xl font-semibold">
              {RESULT_NOT_FOUND}
            </p>
          </td>
        </tr>
      )}
    </tbody>
  );
};

export default memo(TableBody) as typeof TableBody;
