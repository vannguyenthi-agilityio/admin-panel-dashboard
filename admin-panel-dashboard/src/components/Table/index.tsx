export { default as HeaderCell } from './HeaderCell';
export { default as TableHeader } from './TableHeader';
export { default as TableBody } from './TableBody';
export { default as Table } from './Table';
