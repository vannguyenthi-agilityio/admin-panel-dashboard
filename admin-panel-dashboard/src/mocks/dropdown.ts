// Types
import { IListDropDown, IUser } from "@/types";

export const MOCK_USER_DATA: IUser = {
  name: "Derek Alvarado",
  avatar: "https://www.gravatar.com/avatar/2c7d99fe281ecd3bcd65ab915bac6dd5?s=250"
}

export const LIST_DROPDOWN_DATA: IListDropDown[] = [
  {
    href: "/",
    name: "Dashboard"
  },
  {
    href: "/",
    name: "Settings"
  }
]
