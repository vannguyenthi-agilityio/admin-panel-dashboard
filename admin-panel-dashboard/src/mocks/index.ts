export * from './dropdown';
export * from './table';
export * from './form';
