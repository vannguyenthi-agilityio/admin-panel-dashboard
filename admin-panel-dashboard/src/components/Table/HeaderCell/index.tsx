import { memo } from "react";

// Components
import { Button } from "@/components";

// Icons
import { SortIcon } from "@/components/Icons";

export interface IHeaderCell {
  title: string;
  isSortable: boolean;
}

const HeaderCell = ({
  title,
  isSortable,
}: IHeaderCell) => {
  return (
    <div className="flex items-center w-full">
      <div className="text-sm font-normal text-darker">
        {title}
      </div>
      {isSortable && (
        <Button
          buttonType='transprent'
          className='w-[16px] h-[16px] p-0 ml-2'
          aria-label="Arrow Sort"
          leftIcon={<SortIcon />}
        />
      )}
    </div>
  );
};

export default memo(HeaderCell);
