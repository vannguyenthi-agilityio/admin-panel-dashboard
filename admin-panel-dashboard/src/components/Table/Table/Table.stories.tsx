// Libs
import type { Meta, StoryObj } from "@storybook/react";

// Components
import Table from ".";

// Mocks
import { MOCK_CUSTOMER, MOCK_COLUMNS } from "@/mocks";

const meta: Meta<typeof Table> = {
  title: "Components/Table/Table",
  tags: ["autodocs"],
  component: Table,
  argTypes: {
    data: { description: "data of body table" },
    columns: { description: "columns of table" },
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    data: MOCK_CUSTOMER,
    search: {
      field: "id",
      param: "page",
      valueParam: "1",
      placeholder: "Search..."
    },
    columns: MOCK_COLUMNS,
  },
};

export const Empty: Story = {
  args: {
    data: [],
    columns: MOCK_COLUMNS,
  },
};
