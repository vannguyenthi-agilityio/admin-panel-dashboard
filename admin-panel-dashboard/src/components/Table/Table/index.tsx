// Hooks
import {
  memo,
  //useCallback,
} from "react";

// Types
import { IColumnType, ICustomerTable } from "@/types";

// Components
import { TableBody, TableHeader } from "@/components/Table";

interface IPropsTable {
  data: ICustomerTable[];
  columns: IColumnType<ICustomerTable>[];
  pageSize?: number;
  className?: string;
  hasPagination?: boolean;
  total?: number;
  onActionCustomer: (customer: ICustomerTable, action: string) => void;
  onSortFieldCustomer: (key: string) => void;
}

const Table = ({
  data,
  columns,
  className = "",
  onActionCustomer,
  onSortFieldCustomer
}: IPropsTable) => {

  return (
    <div className="w-full">
      <div
        className={`p-0 border-none ring-0 overflow-x-auto ${className}`}>
        <div className="flex flex-col items-start justify-start my-2">
          <table className="w-full" tabIndex={0} id="table">
            <TableHeader columns={columns} onSort={onSortFieldCustomer} />
            <TableBody columns={columns} data={data} onActionCutomer={onActionCustomer}/>
          </table>
        </div>
      </div>
    </div>
  );
};

export default memo(Table);
