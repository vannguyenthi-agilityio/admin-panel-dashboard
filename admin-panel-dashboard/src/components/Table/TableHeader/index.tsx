import { memo } from "react";

// Components
import HeaderCell from "../HeaderCell";

// Types
import { IColumnType, ICustomerTable } from "@/types";

interface IPropsDataTableHeader {
  columns: IColumnType<ICustomerTable>[];
  onSort: (key: string) => void;
}

const TableHeader = ({
  columns,
  onSort,
}: IPropsDataTableHeader) => (
  <thead>
    <tr>
      {columns.map(({ key, title, isSortable = false }) => {
        const handleClickSort = () => {
          isSortable && onSort(key);
        };

        return (
          <th
            data-testid={`header-cell-${title}`}
            onClick={handleClickSort}
            key={key}
            className={`px-6 py-2 leading-[17px] tracking-[0.2px] ${
              isSortable ? "cursor-pointer" : "cursor-default"
            }`}>
            <HeaderCell
              title={title}
              isSortable={isSortable}
            />
          </th>
        );
      })}
    </tr>
  </thead>
);

export default memo(TableHeader);
