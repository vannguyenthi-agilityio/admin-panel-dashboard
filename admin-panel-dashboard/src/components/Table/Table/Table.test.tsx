import { render } from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';

// Mocks
import { MOCK_CUSTOMER, MOCK_COLUMNS } from "@/mocks";

// Components
import Table from ".";

describe("Table", () => {
  const mockActionCutomer = jest.fn();
  const mockProps = {
    data: MOCK_CUSTOMER,
    search: {
      field: "id",
      param: "page",
      valueParam: "1",
      placeholder: "Search..."
    },
    columns: MOCK_COLUMNS,
    onActionCustomer: mockActionCutomer,
    onSortFieldCustomer: mockActionCutomer,
  };

  it("Should matches snapshot", async () => {
    const { container } = render(
      <BrowserRouter>
        <Table {...mockProps} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
