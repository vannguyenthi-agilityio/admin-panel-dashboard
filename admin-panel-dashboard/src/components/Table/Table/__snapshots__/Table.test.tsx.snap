// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Should matches snapshot 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      class="p-0 border-none ring-0 overflow-x-auto "
    >
      <div
        class="flex flex-col items-start justify-start my-2"
      >
        <table
          class="w-full"
          id="table"
          tabindex="0"
        >
          <thead>
            <tr>
              <th
                class="px-6 py-2 leading-[17px] tracking-[0.2px] cursor-default"
                data-testid="header-cell-#"
              >
                <div
                  class="flex items-center w-full"
                >
                  <div
                    class="text-sm font-normal text-darker"
                  >
                    #
                  </div>
                </div>
              </th>
              <th
                class="px-6 py-2 leading-[17px] tracking-[0.2px] cursor-pointer"
                data-testid="header-cell-Full Name"
              >
                <div
                  class="flex items-center w-full"
                >
                  <div
                    class="text-sm font-normal text-darker"
                  >
                    Full Name
                  </div>
                  <button
                    aria-label="Arrow Sort"
                    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-transprent border-none hover:bg-transprent text-sm rounded-lg w-[16px] h-[16px] p-0 ml-2"
                    type="button"
                  >
                    <svg
                      fill="none"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M2.66669 5.16666H13.3334M4.33335 7.83332H11.6667M5.66669 10.5H10.3334"
                        stroke="#6c757d"
                        stroke-linecap="round"
                      />
                    </svg>
                  </button>
                </div>
              </th>
              <th
                class="px-6 py-2 leading-[17px] tracking-[0.2px] cursor-default"
                data-testid="header-cell-Status"
              >
                <div
                  class="flex items-center w-full"
                >
                  <div
                    class="text-sm font-normal text-darker"
                  >
                    Status
                  </div>
                </div>
              </th>
              <th
                class="px-6 py-2 leading-[17px] tracking-[0.2px] cursor-pointer"
                data-testid="header-cell-E-Mail"
              >
                <div
                  class="flex items-center w-full"
                >
                  <div
                    class="text-sm font-normal text-darker"
                  >
                    E-Mail
                  </div>
                  <button
                    aria-label="Arrow Sort"
                    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-transprent border-none hover:bg-transprent text-sm rounded-lg w-[16px] h-[16px] p-0 ml-2"
                    type="button"
                  >
                    <svg
                      fill="none"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M2.66669 5.16666H13.3334M4.33335 7.83332H11.6667M5.66669 10.5H10.3334"
                        stroke="#6c757d"
                        stroke-linecap="round"
                      />
                    </svg>
                  </button>
                </div>
              </th>
              <th
                class="px-6 py-2 leading-[17px] tracking-[0.2px] cursor-pointer"
                data-testid="header-cell-Date of Birth"
              >
                <div
                  class="flex items-center w-full"
                >
                  <div
                    class="text-sm font-normal text-darker"
                  >
                    Date of Birth
                  </div>
                  <button
                    aria-label="Arrow Sort"
                    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-transprent border-none hover:bg-transprent text-sm rounded-lg w-[16px] h-[16px] p-0 ml-2"
                    type="button"
                  >
                    <svg
                      fill="none"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M2.66669 5.16666H13.3334M4.33335 7.83332H11.6667M5.66669 10.5H10.3334"
                        stroke="#6c757d"
                        stroke-linecap="round"
                      />
                    </svg>
                  </button>
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    1
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Alyvia Kelley
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-2.5 w-2.5 rounded-full bg-green-500 me-2"
                  />
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Approved
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    test@gmail.com
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    06/18/1978
                  </p>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    2
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Jaiden Nixon
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-2.5 w-2.5 rounded-full bg-gray-500 me-2"
                  />
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Rejected
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    jaiden.n@gmail.com
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    09/18/1983
                  </p>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    3
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Jaiden Nixon
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-2.5 w-2.5 rounded-full bg-red-500 me-2"
                  />
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    Blocked
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    me@clayton.com
                  </p>
                </div>
              </td>
              <td
                class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
              >
                <div
                  class="flex items-center"
                >
                  <p
                    class="text-darker font-normal text-sm max-w-[200px] truncate"
                  >
                    10/18/1988
                  </p>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
